rfcs/geometric-primitives.md

```diff
@@ -0,0 +1,349 @@
+# Feature Name: `geometric-primitives`
+
+## Summary
+
+Lightweight geometric primitive types for use across bevy engine crates, and as interoperability types for external libraries.
+
+## Motivation
+
+This would provide a standard way to model primitives across the bevy ecosystem to prevent ecosystem fragmentation amongst plugins.
+
+## User-Facing Explanation
+
+Geometric primitives are lightweight representations of geometry that describe the type of geometry as well as fully defined dimensions. These primitives are *not* meshes, but the underlying precise mathematical definition. For example, a circle is:
+
+```rust
+pub struct Circle2d {
+  origin: Point2d,
+  radius: f32,
+}
+```
+
+Geometric primitives have a defined shape, size, position, and orientation. Position and orientation are **not** defined using bevy's `Transform` components. This is because these are fundamental geometric primitives that must be usable and comparable as-is
+
+`bevy_geom` provides two main modules, `geom_2d` and `geom_3d`. Recall that the purpose of this crate is to provide lightweight types, so there are what appear to be duplicates in 2d and 3d, such as `geom_2d::Line2d` and `geom_3d::Line`. Note that the 2d version of a line is lighter weight, and is only defined in 2d. 3d geometry (or 2d with depth which is 3d) is assumed to be the default for most cases. The names of the types were chosen with this in mind, to guide you toward using Line instead of Line2d for example, unless you know why you are making this choice.
+
+## Implementation strategy
+
+### Helper Types
+
+```rust
+/// Stores an angle in radians, and supplies builder functions to prevent errors (from_radians, from_degrees)
```

Comment:
For many applications, I'd prefer to use angles in degrees with a rational rather than float representation. This is essential for things like reversability, commutativity and so on. However, that probably makes more sense as part of another RFC that covers rational versions of all of these primitives (and Bevy's Transforms).

Reply:
To clarify, I don't think that is incompatible with this approach, correct? The goal is to make it obvious through the API what kind of "angle" is being supplied. I've messed up deg/rad enough times to see the value in type/API-level guardrails.

Along the lines of your suggestion, it may also make sense to constrain the underlying angle to be 0-2pi, with a counter for n_revolutions if needed. This would ensure 12deg == 372deg == 732deg without losing angular precision as the angle grows.